### PR TITLE
Feature/transformable objects

### PR DESCRIPTION
**Pull request by toloudis**
This code change accomplishes the following:
1. begin to introduce an object hierarchy so that scene objects can be associated with transforms.  e.g. `SceneObject` is a container for a transform.  `SceneLight` is a `SceneObject` (which has a transform) that has a Light.
2. manipulators can operate on those transforms
3. fix several bugs with the rotation tool
4. split large files into separate files per class (AreaLightTool, MoveTool, RotateTool etc)
5. fix one floating-point shading bug on Windows in the gui shader
6. Area lights are currently manipulated by rotation of their direction vector only (as seen in `SceneLight::updateTransform`)

### CODE REVIEW

File: renderlib/Light.h [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    m_InvHeight = Other.m_InvHeight;
+    m_HalfHeight = Other.m_HalfHeight;
+    m_InvHalfHeight = Other.m_InvHalfHeight;
+    m_Distance = Other.m_Distance;
+    m_SkyRadius = Other.m_SkyRadius;
+    m_P = Other.m_P;
+    m_Target = Other.m_Target;
+    m_N = Other.m_N;
+    m_U = Other.m_U;
+    m_V = Other.m_V;
```
Comment (ShrimpCryptid) [inline renderlib/Light.h:94]: Do glm::vec3 values need to be copied? Not sure what their behavior is for equality operator

File: renderlib/Origins.cpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+#include "Logging.h"
+#include "RenderSettings.h"
+
+void
+Origins::update(SceneView& scene)
+{
+  if (scene.scene) {
+
+    // e.g. find all selected objects in scene and collect up their centers/transforms here.
+    SceneLight& lt = scene.scene->m_lighting.m_sceneLights[1];
```
Comment (ShrimpCryptid) [inline renderlib/Origins.cpp:13]: Is this a TODO?

File: renderlib/Light.h (hunk 1 above)
Comment (toloudis) [reply renderlib/Light.h:94]: In c++ you can generally assume that assignment is a copy.

File: renderlib/Origins.cpp (hunk 2 above)
Comment (toloudis) [reply renderlib/Origins.cpp:13]: yes, once the framework is completely adopted / generalized, this code won't have to access the area light in this way - it will go through something like scene.selectedObject().

File: renderlib/AreaLightTool.cpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+  gesture.graphics.addCommand(GL_LINES);
+  gesture.graphics.addLine(Gesture::Graphics::VertsCode(p + v0 * 0.1f, color, opacity, code),
+                           Gesture::Graphics::VertsCode(t + v0 * 0.1f, color, opacity, code));
+  gesture.graphics.addLine(Gesture::Graphics::VertsCode(p + v1 * 0.1f, color, opacity, code),
+                           Gesture::Graphics::VertsCode(t + v1 * 0.1f, color, opacity, code));
+  gesture.graphics.addLine(Gesture::Graphics::VertsCode(p + v2 * 0.1f, color, opacity, code),
+                           Gesture::Graphics::VertsCode(t + v2 * 0.1f, color, opacity, code));
+  gesture.graphics.addLine(Gesture::Graphics::VertsCode(p + v3 * 0.1f, color, opacity, code),
+                           Gesture::Graphics::VertsCode(t + v3 * 0.1f, color, opacity, code));
```
Comment (ShrimpCryptid) [inline renderlib/AreaLightTool.cpp:51]: This code is drawing lines from the light towards the target, right? Is there a reason for it not to be grouped with the existing line drawings above? (redundant call to `gesture.graphics.addCommand(GL_LINES)`?)

Also might be nice to add a comment for each of these blocks to say what they're drawing

File: renderlib/MathUtil.cpp [hunk 4]
```diff
@@ -33,3 +33,11 @@ AffineSpace3f::inverse() const
...
   return AffineSpace3f(il, ip);
 }
+
+AffineSpace3f::AffineSpace3f(const glm::quat& orientation, const glm::vec3& p)
+  : p(p)
+{
+  // convert quat to linspace:
+  glm::mat3 m = glm::mat3_cast(orientation);
+  l = LinearSpace3f(m[0], m[1], m[2]);
+}
```
Comment (ShrimpCryptid) [inline renderlib/MathUtil.cpp:43]: Does `l`/`LinearSpace3f` describe a rotation matrix?

File: renderlib/RotateTool.cpp [hunk 5]
```diff
@@ -134,45 +134,45 @@ RotateTool::action(SceneView& scene, Gesture& gesture)
...
+        glm::vec3 vN = camFrame.vz;
+        angle = getDraggedAngle(camFrame.vz, p, l, l0, l1);
         motion = glm::angleAxis(angle, vN);
       } break;
       case RotateTool::kRotate: // general tumble rotation
         // use camera trackball algorithm
         // scale pixels to radians of rotation (TODO)
         float xRadians = -button.drag.x * dragScale * glm::two_pi<float>();
-        float yRadians = -button.drag.y * dragScale * glm::two_pi<float>();
+        float yRadians = button.drag.y * dragScale * glm::two_pi<float>();
```
Comment (ShrimpCryptid) [inline renderlib/RotateTool.cpp:175]: There's another place in the code where a y-component also needs to be negated, is this related?

File: renderlib/RotateTool.cpp [hunk 6]
```diff
@@ -227,13 +228,22 @@ RotateTool::draw(SceneView& scene, Gesture& gesture)
...
   // Draw the manipulator to be at some constant size on screen
   float scale = length(viewDir) * scene.camera.getHalfHorizontalAperture() * (s_manipulatorSize / resolution.x);
 
   AffineSpace3f axis;
   axis.p = target.p;
 
+  // TODO allow switching between local and world coordinate space!
+  // axis.l.vx = m_rotation * target.l.vx; // glm::vec3(1, 0, 0);
+  // axis.l.vy = m_rotation * target.l.vy; // glm::vec3(0, 1, 0);
+  // axis.l.vz = m_rotation * target.l.vz; // glm::vec3(0, 0, 1);
```
Comment (ShrimpCryptid) [inline renderlib/RotateTool.cpp:242]: ooh that will be nice

File: renderlib/graphics/glsl/GLGuiShader.cpp [hunk 7]
```diff
@@ -46,18 +46,19 @@ static const char* fragment_shader_text =
...
-            result.w = 1.0;
+        if (picking == 1) {
+          result.w = 1.0;
+        }
 
         // Gesture geometry handshake: any uv value below -64 means
         // no texture lookup. Check VertsCode::k_noTexture
-        if (picking == 0 && Frag_UV.s > -64)
-            result *= texture(Texture, Frag_UV.st);
+        if (picking == 0 && Frag_UV.x > -63.0) {
```
Comment (ShrimpCryptid) [inline renderlib/graphics/glsl/GLGuiShader.cpp:55]: Should this be `Frag_UV.x >= -63`?

File: renderlib/Object3d.h [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+#include "MathUtil.h"
+
+// an object that can be transformed in 3d space
+class Transform3d
+{
+public:
+  Transform3d()
+  {
+    m_rotation = glm::angleAxis(0.0f, glm::vec3(0.0f, 0.0f, 1.0f));
```
Comment (ShrimpCryptid) [inline renderlib/Object3d.h:11]: Is there a difference between using `angleAxis` and `glm::quat` here?

**Review by ShrimpCryptid — APPROVED** [review]
Looks good! Scattered some comments and questions in but nothing urgent.

File: renderlib/MathUtil.cpp (hunk 4 above)
Comment (toloudis) [reply renderlib/MathUtil.cpp:43]: yes, basically it's a 3x3 matrix (or set of 3 basis vectors) but it doesn't strictly have to be a rotation.  In all cases I'm using it, it is a rotation e.g. between camera and world space

File: renderlib/RotateTool.cpp (hunk 5 above)
Comment (toloudis) [reply renderlib/RotateTool.cpp:175]: Yeah - I'll look for it and see if they can be factored together

File: renderlib/RotateTool.cpp (hunk 6 above)
Comment (toloudis) [reply renderlib/RotateTool.cpp:242]: next PR will implement this

File: renderlib/graphics/glsl/GLGuiShader.cpp (hunk 7 above)
Comment (toloudis) [reply renderlib/graphics/glsl/GLGuiShader.cpp:55]: I'll add a comment here.  This was fixing a windows-only (and maybe nvidia-only) bug where the shader could not figure out whether `-64 > -64` .  The lines drawn were choppy.  So basically this is really Frag_UV.x >= (-64+epsilon)

File: renderlib/Object3d.h (hunk 8 above)
Comment (toloudis) [reply renderlib/Object3d.h:11]: I'll switch it to the simpler version that initializes it with a vec3 of euler angles.

File: renderlib/CCamera.cpp [hunk 9]
```diff
@@ -238,7 +208,11 @@ cameraManipulation(const glm::vec2 viewportSize, Gesture& gesture, CCamera& came
     glm::vec3 rotated_v = camera.m_From - camera.m_Target;
     glm::vec3 v = camera.m_From - camera.m_Target;
     glm::vec3 newEye, newUp;
-    trackball(drag.x, -drag.y, camera, newEye, newUp);
+
+    glm::quat q = trackball(drag.x, -drag.y, camera.m_From - camera.m_Target, camera.m_Up, camera.m_U);
+    newUp = q * camera.m_Up;
+    newEye = q * rotated_v + camera.m_Target;
+
     rotated_v = newEye - camera.m_Target;
```
Comment (ShrimpCryptid) [inline renderlib/CCamera.cpp:216]: Does adding and then removing `camera.m_Target` do anything here?

Comment (toloudis) [reply renderlib/CCamera.cpp:216]: nope! I wondered if anyone would notice 😁

Comment (toloudis) [reply renderlib/CCamera.cpp]: This is now resolved in one hopefully last commit on this PR
